File: codeforces/1553-G/submissions/OK/127331064.py
n = 1e6+5
a, wei, p = [0] * n, [0] * n, [0] * n
f, color, d = [0] * n, [0] * n, [0] * n

def lower_bound(lst, val):
    return lst[val]


def unique(lst):
    return lst


def readline():
    return map(int, input().split())


def find(x):
    if x == f[x]:
        return x
    f[x] = find(f[x])
    return f[x]


def main():
    top = 0
    k1 = [0] * 1000000
    nn, q = readline()
    a = list(readline())
    for i in range(1, nn + 1):
        a[i] = readline()
        f[i] = i
        wei[a[i]] = i

    for i in range(2, n):
        if p[i]:
            continue
        pp = -1
        for j in range(i, n, i):
            if wei[j]:
                if pp == -1:
                    color[i] = find(wei[j])
                else:
                    f[find(pp)] = find(wei[j])
                pp = wei[j]
            p[j] = i

    for i in range(2, n):
        if p[i] == i:
            color[i] = find(color[i])

    c = 0

    for i in range(1, nn + 1):
        az = a[i] + 1
        c += 1
        d[c] = find(i)

        while az > 1:
            c += 1
            d[c] = color[p[az]]
            az /= p[az]

        d[1:c + 1] = sorted(d[1:c + 1])
        c = unique(d[1: c + 1]) - d[0] - 1

        for k in range(1, c):
            for j in range(k + 1, c + 1):
                top += 1
                k1[top] = (d[k], d[j])

        c = 0

    k1[1:top + 1] = sorted(k1[1:top + 1])
    for i in range(1, q + 1):
        x, y = readline()
        x = find(x)
        y = find(y)
        if x == y:
            print(0)
        else:
            ask = (min(x, y), max(x, y))
            jz = lower_bound(k1[1: top + 1], ask)
            if 1 <= jz and jz <= top and k1[jz] == ask:
                print(1)
            else:
                print(2)


if __name__ == '__main__':
    main()

File: codeforces/1553-G/submissions/OK/127331064.cpp
#include<bits/stdc++.h>
#define N 1000005
using namespace std;
inline int read(){
 int x=0,f=1;
 char ch=getchar();
 while(ch<'0' || ch>'9'){if(ch=='-')f=-1;ch=getchar();}
 while(ch>='0'&&ch<='9')x=(x<<3)+(x<<1)+(ch^48),ch=getchar();
 return x*f;
}
int n,q,a[N],t[N],pr[N];
int f[N],col[N],d[N],c;
int find(int x){
 return f[x]==x?x:f[x]=find(f[x]);
}
vector <pair<int,int> > res;
signed main(){
	n=read(),q=read();
 for(int i=1;i<=n;i++) a[i]=read(),f[i]=i,t[a[i]]=i;
 for(int i=2;i<N;i++){
 if(pr[i])continue;
 for(int j=i,pre=-1;j<N;j+=i){
 if(t[j]){
 if(pre==-1)col[i]=find(t[j]);
 else f[find(pre)]=find(t[j]);
				pre=t[j];
 } pr[j]=i;
 }
 }
 // for(int i=1;i<=n;++i) cout<<find(i)<<' ';
 // cout<<endl;
 for(int i=2;i<N;i++)if(pr[i]==i)col[i]=find(col[i]);
 for(int i=1,p=a[1]+1;i<=n;p=a[++i]+1,c=0){
		d[++c]=find(i);
 while(p>1) d[++c]=col[pr[p]],p/=pr[p];
		sort(d+1,d+c+1);
		c=unique(d+1,d+c+1)-d-1;
 for(int i=1;i<c;i++){
 for(int j=i+1;j<=c;j++){
				res.push_back({d[i],d[j]});
 }
 }
 }
	sort(res.begin(),res.end());
 for(int i=1;i<=q;i++){
 int x=read(),y=read();
		x=find(x),y=find(y);
 if(x==y)puts("0");
 else{
			pair<int,int> cur={min(x,y),max(x,y)};
 auto it=lower_bound(res.begin(),res.end(),cur);
			puts(it!=res.end()&&*it==cur?"1":"2");
 // cout<<(*it).first<<' '<<(*it).second<<endl;
 } 
 }
 return 0;
}
